packages/spacecat-shared-gpt-client/src/clients/firefall-client.js

```diff
@@ -161,23 +163,113 @@ export default class FirefallClient {
     return jobStatusResponse;
   }
 
-  async fetch(prompt) {
+  /**
+   * Fetches data from Firefall Chat Completion API.
+   * @param prompt The text prompt to provide to Firefall
+   * @param options The options for the call:
+   *                - imageUrls: An array of URLs of the images to provide to Firefall
+   * @returns {Promise<*>}
+   */
+  async fetchChatCompletion(prompt, options = {}) {
     if (!hasText(prompt)) {
       throw new Error('Invalid prompt received');
     }
+    const hasImageUrls = options?.imageUrls && options.imageUrls.length > 0;
 
     try {
       const startTime = process.hrtime.bigint();
-      const jobSubmissionResponse = await this.#submitJob(prompt);
-      const jobStatusResponse = await this.#pollJobStatus(jobSubmissionResponse.job_id);
-      this.#logDuration('Firefall API call', startTime);
+      const userRole = {
+        role: 'user',
+        content: [
+          {
+            type: 'text',
+            text: prompt,
+          },
+        ],
+      };
+      const body = {
+        llm_metadata: {
+          model_name: this.config.FIREFALL_API_CAPABILITY_NAME,
+          llm_type: 'azure_chat_openai',
+        },
+        messages: [
+          {
+            role: 'system',
+            content: 'You are a helpful assistant designed to output JSON.',
+          },
+          userRole,
+        ],
+      };
+      if (hasImageUrls) {
+        if (!Array.isArray(options?.imageUrls)) {
+          throw new Error('imageUrls must be an array.');
+        }
+        // Warn if model might not handle images.
+        if (!this.config.capabilityName.includes('vision') && this.log?.warn) {
```

Comment:
If images are provided can we just override the model name to always be gpt-4-vision? How much knowledge of Open AI models do we want users of the client to understand vs just letting the client figure out the best model to use?

Reply:
Nice.  I wasn't sure about this check.  Without the right model, the images are ignored.  But I didn't want to lock us into specific models, as they'll change (maybe one will come out using `images` instead of `vision`).  Allowing any model and allowing Firefall to reject them would be more future proof.
Or default to `"gpt-4-turbo"` and switch to `"gpt-4-vision"` if there are images?  Any allowances to use other models to use fewer credits or more advanced responses?
What do you think?

Reply:
If you want to stay open ended and less opinionated then you probably should do any validation then since you won't know what model name to validate against. If image urls are provided just add them to the messsages payload and pss it along to Firefall.